Introducing delegate if there is GraphQL errors

kimdv opened the pull request:
The idea of this delegate is based on 
- https://github.com/apollographql/apollo-link/tree/master/packages/apollo-link-error
- https://blog.apollographql.com/full-stack-error-handling-with-graphql-apollo-5c12da407210

It adds ability for better error handling on the clients when there is Graphql errors  
Etc when a token is expired. 

The diffrent from this delegate and `HTTPNetworkTransportRetryDelegate` is that some errors might come from the GraphQL API and not from middleware etc. 

 - [x] Add tests

designatednerd commented on Sources/Apollo/HTTPNetworkTransport.swift:
```diff
@@ -167,6 +183,25 @@ public class HTTPNetworkTransport {
+  private func handleGraphQLErrorsOrComplete<Operation>(operation: Operation,
+                                                        response: GraphQLResponse<Operation>,
+                                                        completionHandler: @escaping (_ result: Result<GraphQLResponse<Operation>, Error>) -> Void) throws {
+    if let delegate = self.delegate as? NetworkGraphQLErrorDelegate,
```
Personally I'd `guard let` this for clearer control flow.  

You may also want to look at some of the changes proposed in #767 - there's some stuff in there that's similar to this just without the delegate, including something to only parse the errors (rather than parsing the entire data structure as you're doing below).

kimdv replied:
Added `guard let`

Related to the #767 are you thinking of the `func parseErrorsFast()`?

designatednerd replied:
yep!

designatednerd commented on Sources/Apollo/HTTPNetworkTransport.swift:
```diff
@@ -167,6 +183,25 @@ public class HTTPNetworkTransport {
     return task
   }
+
+  private func handleGraphQLErrorsOrComplete<Operation>(operation: Operation,
```
I think this needs to include a `files` parameter in case an operation with files needs to get retried

kimdv replied:
Just saw the `handleErrorOrRetry` also misses that? 
Want me to add here or in a new PR?

designatednerd replied:
Add here - I think the reason it doesn't handle it there is that it's more specifically looking at retries for APQ stuff rather than in general, and APQ can only be done with Queries, not mutations.

kimdv commented:
Will create tests when we can mock. Depending on #767

designatednerd commented:
@kimdv #767 is finally merged if you wanna use the stuff from that still

designatednerd commented on Sources/Apollo/HTTPNetworkTransport.swift:
```diff
@@ -131,10 +146,11 @@ public class HTTPNetworkTransport {
+                                 error: receivedError,
+                                 for: request,
+                                 response: response,
+                                 completionHandler: completionHandler)
```
I think the indentation is off here and below

designatednerd commented on Sources/Apollo/HTTPNetworkTransport.swift:
```diff
@@ -183,17 +203,47 @@ public class HTTPNetworkTransport {
+                                                        completionHandler: @escaping (_ result: Result<GraphQLResponse<Operation>, Error>) -> Void) throws {
+    guard let delegate = self.delegate as? HTTPNetworkTransportGraphQLErrorDelegate,
+      let graphQLErrors = response.parseErrorsOnlyFast(),
+      !graphQLErrors.isEmpty else {
```
this is more of a note to self than to you: I need to add my readability `isNotEmpty` boolean extension to this code 🙃

designatednerd replied:
Done in #872

kimdv commented:
@designatednerd I rebased and added two tests. One with error and one with none.

kimdv commented on Tests/ApolloTests/HTTPTransportTests.swift:
```diff
@@ -193,6 +196,50 @@ class HTTPTransportTests: XCTestCase {
+  func testErrorDelegateWithErrors() throws {
+    self.graphQlErrors = []
+    let query = HeroNameQuery()
+    let body = ["errors": [["message": "Test graphql error"]]]
```
Because `GraphQLError` is not encodable or NSObject we can not use it. 
Maybe we should consider to make `GraphQLError` encodable?

kimdv commented on Tests/ApolloTestSupport/MockURLSession.swift:
```diff
@@ -9,19 +9,25 @@ import Foundation
+
+  public var data: Data?
+  public var response: URLResponse?
+  public var error: Error?
```
This way we can handle what the completion should return

kimdv commented:
@designatednerd I rebased and added two tests. One with error and one with none. 
I'm not sure that the tests are good enough. If not, please send some hints! 🚀

designatednerd commented on Tests/ApolloTests/HTTPTransportTests.swift:
```diff
@@ -266,3 +313,11 @@ extension HTTPTransportTests: HTTPNetworkTransportRetryDelegate {
+
+extension HTTPTransportTests: HTTPNetworkTransportGraphQLErrorDelegate {
+  func networkTransport(_ networkTransport: HTTPNetworkTransport, receivedGraphQLErrors errors: [GraphQLError], retryHandler: @escaping (Bool) -> Void) {
+    self.graphQlErrors = errors
```
doesn't the retry handler need to be called here?

kimdv replied:
Doh 🤦‍♂️
Will be fixed

designatednerd commented:
@kimdv can you merge in and/or rebase off `master` so the test suite can run on Circle please?

kimdv commented:
@designatednerd done! 
Let's see what explodes 🔥🙈

designatednerd commented:
🎉 Nothing! 🎉 Mergin'

RolandasRazma commented:
how do you know if failure after retrying? 
if request fails and you call `retryHandler(true)` and it fails again I don't see how to know that you should just give up

designatednerd commented:
That's why we explicitly recommend against just calling `retryHandler(true)` without any context in the documentation for this method and the retry delegate for network errors. 

If you're not looking at the actual errors that have come in, you could easily wind up in an infinite retry loop.

RolandasRazma commented:
I am. I'm thinking about case where server say "token invalid", I update token, and server say "still invalid". Correct thing to do is to give up, but I don't know that I already tried to update token

designatednerd commented:
It sounds a bit more like a bug in your server if a just-updated token is being rejected as invalid, but you could theoretically add some tracking of what the last (n) errors received were, and if some aspect of them (status code, message, etc) is the same, then stop retrying.

RolandasRazma commented:
maybe it's me, but I like to code defensively 🤷‍♂

designatednerd commented:
That's fair, but I think the "how do I handle situation x" stuff is more of an application-level decision, so we're letting developers handle that as defensively as they'd like to.

kimdv commented:
Hi! 
We had the same problem but solved it in the client. 
``` swift
final class GraphQLClient {
    private var retryHandlers = [(Bool) -> Void]()
    private let isolationQueue = DispatchQueue(label: "GraphQLClient.isolationQueue",
                                               qos: .userInitiated,
                                               attributes: [],
                                               autoreleaseFrequency: .workItem,
                                               target: nil)
}
```

``` swift
extension GraphQLClient: HTTPNetworkTransportGraphQLErrorDelegate {
    func networkTransport(_ networkTransport: HTTPNetworkTransport, receivedGraphQLErrors errors: [GraphQLError], retryHandler: @escaping (Bool) -> Void) {
        let errors = errors.map { OurGraphQLError(error: $0) }
        
        guard error.contains(.authTokenInvalid) else {
            retryHandler(false)
            return
        }
        
        let isTokenExchangeInProgress = scheduleForExecution(retryHandler: retryHandler)
        if isTokenExchangeInProgress { return }
        
        exchangeRefreshToken { [weak self] result in
            guard let strongSelf = self else { return }
            
            switch result {
            case .success:
                strongSelf.executeRetryHandlers(retry: true)
            
            case .failure(let error):              
                strongSelf.executeRetryHandlers(retry: false)
            }
        }
    }

    private func scheduleForExecution(retryHandler: @escaping (Bool) -> Void ) -> Bool {
        var isTokenExchangeInProgress = true
        isolationQueue.sync {
            isTokenExchangeInProgress = !retryHandlers.isEmpty
            retryHandlers.append(retryHandler)
        }
        return isTokenExchangeInProgress
    }

    private func executeRetryHandlers(retry: Bool) {
        isolationQueue.async { [weak self] in
            guard let strongSelf = self else { return }
            for handler in strongSelf.retryHandlers {
                handler(retry)
            }
            strongSelf.retryHandlers = []
        }
    }
}
```

In this way we can cancel if refresh token exchange fails.
